### DISCUSSION TITLE
how to use replacements when using PowerShell in Windows

### DISCUSSION THREAD

**Discussion by huangjj27** [BurntSushi/ripgrep#2215] [Q&A] [answered]
#### What version of ripgrep are you using?
```
> rg --version
ripgrep 13.0.0
-SIMD -AVX (compiled)
+SIMD +AVX (runtime)
```
#### How did you install ripgrep?
`cargo install ripgrep`

#### What operating system are you using ripgrep on?

Windows 10

#### What are the steps to reproduce the behavior?

here is the file.txt: 

```
10a3789
51b7883
```

and then I run in PowerShell:
```
rg -N "^(\w+)$" file.txt -r "'$1',"
```

#### What is the actual behavior?

I get these lines:
```
'',
'',
```

but without the replacement, it captured:
```
> rg -N "^(\w+)$" file.txt                   
10a3789
51b7883
```
#### What is the expected behavior?
as I want to add a single quote both before and after each line, and then add a comma to end each line expecting it cat these lines:
```
'10a3789',
'51b7883',
```

**Comment by BurntSushi** [comment]
I can reproduce your problem:

```
$ rg -N '^(\w+)$' /tmp/file.txt -r "'$1',"
'',
'',
```

But this looks like a shell problem, not a ripgrep problem. In particular, single quotes have nothing to do with this. If you remove your single quotes, you should get the same behavior. In my shell, I have to use `\$` to avoid having it interpreted as an environment variable:

```
$ rg -N '^(\w+)$' /tmp/file.txt -r "'\$1',"
'10a3789',
'51b7883',
```

Using just single quotes is possible too, although it is a bit unwieldy and I have no idea whether it works on Windows or not:

```
$ rg -N '^(\w+)$' /tmp/file.txt -r "'"'$1'"'"','
'10a3789',
'51b7883',
```

**Comment by huangjj27** [comment, marked as answer]
It seems neither of the commands works on Powershell:
```
> rg -N '^(\w+)$' file.txt -r "'\$1',"          
'\',
'\',
> rg -N '^(\w+)$' file.txt -r "'"'$1'"'"','  
$1: The system cannot find the file specified. (os error 2)
': The system cannot find the file specified. (os error 2)
,: The system cannot find the file specified. (os error 2)
file.txt
'
'
```

and I tried both of the commands in my WSL, both works:
```
$ rg -N '^(\w+)$' file.txt -r "'\$1',"
'10a3789',
'51b7883',
$ rg -N '^(\w+)$' file.txt -r "'"'$1'"'"','
'10a3789',
'51b7883',
```

hope these information will help others

**Reply by BillHally** [reply]
PowerShell uses a backtick character to prevent variable substitution in double-quoted strings, so the following will work:

```powershell
rg -N '^(\w+)$' file.txt -r "'`$1',"
```

**Comment by huangjj27** [comment]
Here is another example. if my replacement string quoted by double quote marks, captures are missing:
```
> echo "2012-03-14, 2013-01-01 and 2014-07-05" | rg '(?P<y>\d{4})-(?P<m>\d{2})-(?P<d>\d{2})' --replace "$m/$d/$y"
//, // and //
```

but if they are quoted by single quote marks, variables are captured:
```
> echo "2012-03-14, 2013-01-01 and 2014-07-05" | rg '(?P<y>\d{4})-(?P<m>\d{2})-(?P<d>\d{2})' --replace '$m/$d/$y'
03/14/2012, 01/01/2013 and 07/05/2014
```

This may be a helpful walkaround when we don't contain single quote marks `'` in our replacement.
